### ISSUE TITLE
Client may hang after long-time pressure test

### ISSUE THREAD

**Issue by young-mu** [Azure/azure-iot-sdk-node#186]
# Context
- **OS and version used:** MacOS Sierra 10.12.6
- **Node.js version:** v8.1.2
- **npm version:** 5.0.3
- **list of installed packages:** (too long, just dependencies of **azure-iot-device-mqtt**)

# Description of the issue:
I used **azure-iot-device-mqtt** module to do pressure test. I found after long time (about 2 or 3 hours) the client will hang, does NOT enter the callback of *sendEvent*,  also does NOT emit any events like *error* or *disconnect*. This issue is random. I once encountered few times. 

# Code sample exhibiting the issue:

```javascript
var Protocol = require('azure-iot-device-mqtt').Mqtt;
var Client = require('azure-iot-device').Client;
var Message = require('azure-iot-device').Message;

var connectionString = '<connectionString>'; // I have removed mine

var client = Client.fromConnectionString(connectionString, Protocol);

var connectCallback = function (err) {
    if (err) {
        console.log('connect failed with error', err);
        return;
    }

    console.log('connect successfully');

    var i = 0;
    var sendInterval = setInterval(function () {
        var message = new Message('I am device @ ' + i);
        client.sendEvent(message, function (err) {
            if (err) {
                console.log('send failed with error', err);
                return;
            }
            console.log('send successfully @ ' + i + ' (' + new Date() + ')');
            i++;
        });
    }, 10000);

    client.on('message', function (msg) {
        console.log('message: ' + msg.data.toString());
    });

    client.on('error', function (err) {
        console.log('error', err.message);
    });

    client.on('disconnect', function () {
        console.log('disconnect');
        clearInterval(sendInterval);
        client.removeAllListeners();
        client.open(connectCallback);
    });
};

client.open(connectCallback);
```

**Comment by pierreca** [comment]
"pressure test" to do what? how many messages were you sending over what period? not that you're doing anything wrong but i'm trying to understand what code was running. I see the sample piece of code sends a message every 10 seconds, but how many of those did you have running in parallel that make you call this "pressure test"?

also if you could share your hub name and the time at which the issue happened i can try and see if there is anything in the logs - but as is, i don't have enough information to weigh in on this issue.

**Comment by young-mu** [comment]
The "pressure test" here just means that I want to test the SDK stability for a long time. Maybe this word is not correct. Anyway, just as what I have described above, `after 2 or 3 hours, the client will hang, does NOT enter the callback of sendEvent, also does NOT emit any events like error or disconnect.`

The following is a screenshot of my test. Now is 14:37, but the last message is 14:26.

The hostname is **RuffTest2.azure-devices.cn** and deviceId is **ruff_node**.

![1514443047134](https://user-images.githubusercontent.com/7543236/34402480-bfd38012-ebdc-11e7-9f66-735501b69bbc.jpg)

**Comment by pierreca** [comment]
thanks for the added details, I'll look into it. can you let me know what version of the SDK packages you are using?

also, how reliable is the network connection here? not that I blame it, the SDK should handle whatever connection is thrown at it but I'm still trying to understand if there are special circumstances surrounding the apparent loss of connectivity.

**Comment by young-mu** [comment]
The package `azure-iot-device-mqtt` version is 1.2.3. 

I use office network, it should be reliable.

**Comment by pierreca** [comment]
@young-mu just realized the issue happens on MacOS X. socket disconnections are handled kinda differently from one OS to the next, I'm not sure how Mac OS works in that respect. Just curious does it repro on the device OS itself or is that code meant to be run on the mac?

**Comment by pierreca** [comment]
@sharathbaddam I appreciate you chiming on this but a lot more information is needed for us to do anything useful with this.
- what OS is the code running on?
- what protocol/transport package are you using?
- what version of the SDK packages?
- does this repro consistently?

also just to confirm:
- when you say the connection is resumed, do you mean the messages are sent? or that the new messages are sent but not the old ones whose callback is never called? or that the network connection comes back up but messages do not get sent?

**Comment by sharathbaddam** [comment]
@pierreca here are the details.

- _what OS is the code running on?_
We are running on Windows 10 with node 6.9.5

- _what protocol/transport package are you using?_
MQTT

- _what version of the SDK packages?_
azure-iothub@1.2.2, azure-iot-device-mqtt@1.2.3

- _does this repro consistently?_
It repoduces consistently but not predictably. Sometimes it happens every few minutes, other times after a few days. Eventually it seems that it will always occur.

also just to confirm:

- when you say the connection is resumed, do you mean the messages are sent? or that the new messages are sent but not the old ones whose callback is never called? or that the network connection comes back up but messages do not get sent?

In this situation that the callback is not called, we force a timeout of about 150 seconds on our end to move on. It does appear that sometimes data is sent even when the callback is not being called, but we do not know if this is always the case.
Our situation is :- streaming data is coming into our system and we call sendEvent at some periodic rate. After 150 seconds, if the callback is not called, we move on. Even while we are waiting for the callback on previous call,  sendEvent function is called with new data at same periodic rate. So sometimes we don't see callback getting called.
This behaviour does not only happen when we have a network disconnection from disconnecting ethernet cable, it happens just during normal running.

**Comment by pierreca** [comment]
@sharathbaddam Do you configure the retry parameters/policy at all? 150 seconds is shorter than our maximum retry timeout so if you leave the defaults, and the SDK is still retrying, it's normal not to get answers. What I don't understand if why would the SDK be retrying if everything's fine like you describe.

Just to be clear: we have a comprehensive set of end-to-end, fault injection and long-haul tests running on all protocols so while we're fully aware there might be bugs, it's simply extremely rare that the code malfunctions to the extent you describe. so I *need* to understand as much as possible your environment to figure out the repro steps. You better believe we're not shipping stuff we know to be unstable.

**Comment by njmarang** [comment]
@pierreca, just to add to the above.  If you could provide us with the information for setting retry and timeouts on your end, please let us know.  We were not aware that there was one when using MQTT.

We have similar tests in place to the ones that you are describing.  From our point of view, calling sendEvent is pretty straightforward, and there is a callback that we expect to be called.  Again, we wait 150 seconds, which is arbitrary, and could be changed - but we need some indicator that we have not received success/failure so that we can move on with other data or move the data somewhere else.

In the situations that it happens, we have seen data go through - the callback simply does not get called, just as the first person that started this issue has indicated.  So it doesn't appear to have anything to do with retrying.  The data does seem to post, the callback simply does not get called.

For us, this has happened across all systems we have deployed this package.  We produce it literally in every instance of this that we have running.  The system characteristics are as @sharathbaddam described above, but we have tried with multiple different computers/networks/servers.  All the same result.

So to your point, we are under the same pressure from others to explain why they are experiencing this issue on their deployed systems.

For example, pulling the ethernet cable did not result in errors on the callback function like we expected.  Otherwise, there seems to be no anomalous behavior that triggers this scenario.

Our code looks similar to the code posted above:

```javascript
this.postClient.sendEvent(clientmessage, (error, result) => {
    if (!error) {
        res();
    }
    else {
        rej(error);
    }
});
```

**Comment by anthonyvercolano** [comment]
Any chance you could try this using amqp?

**Comment by pierreca** [comment]
@njmarang @sharathbaddam The way we do retries is described here: 
https://github.com/Azure/azure-iot-sdk-node/wiki/Connectivity-and-Retries

Please let us know if that helps or if you'd like to see additional things in this document?

Now onto trying to solve the problem or at least reduce the potential causes:
If you'd rather implement your own retry logic or want to disable it alltogether, you could go 2 ways:
1. Create a class that implements the `RetryPolicy` interface and use `client.setRetryPolicy()` to set it.
2. Disable retries by calling `client.setRetryPolicy(new NoRetry())` (the NoRetry class is defined in the `azure-iot-common` package). 

If you have the environment to reproduce the issue pretty consistently, you can also set the `DEBUG` environment variable to `*` (to see everything) or at least to `azure-iot*` which should log the states and actions of the transport layers and see if the state machine gets stuck in a particular state. That would greatly help narrow down where the problem is...

**Comment by njmarang** [comment]
Thank you @anthonyvercolano and @pierreca.  We will look into amqp and retryPolicy details.  Just a note - we have updated to your latest version and ran through the weekend without seeing this issue.  I don't know if you have any comment on 1.x versus 2.x in this regard.

**Comment by njmarang** [comment]
OK It just now happened with your latest version also....it made it almost a week, so that's pretty good I guess.

**Comment by sharathbaddam** [comment]
Yes it is still using mqtt, that was the test we started last week.

**Comment by pierreca** [comment]
quick update to point out that we're investigating something that may be related...

https://github.com/mqttjs/MQTT.js/pull/787

**Comment by sharathbaddam** [comment]
@pierreca @anthonyvercolano Can we change the default timeout for iot-hub? is it configurable ?

**Comment by pierreca** [comment]
@sharathbaddam Are you asking about the default timeout for the SDK retry logic or another timeout that is handled server-side? or a particular keep-alive/protocol specific timeout?

**Comment by njmarang** [comment]
@pierreca thank you for all the support and checking in.  We have updated all latest packages, grabbed the MQTT stuff, and extended the timeout on our side (our code) so that we don't timeout before expected timeout of the underlying libraries. 

We have been running for some time now and have not seen any issues.  I think we are considering this closed on our end.

**Comment by pierreca** [comment]
@njmarang thanks! we'll keep watching for this type of issues with our long haul tests.
@sharathbaddam @young-mu feel free to reopen if you still see the problem or even better, open a new issue with additional information.

*Please help us improve by filling out this 2-minute [customer satisfaction survey](https://www.surveymonkey.com/r/Q5PLZ3V).*
